File: src/Day2/Part1.hs
module Day2.Part1 where

import Day2.Util ( Entry (..), parseEntry )

day2part1 :: String -> String
day2part1 input = show $ solve entries
  where entries = map parseEntry $ lines input

solve :: [Entry] -> Int
solve = length . filter isValid

isValid :: Entry -> Bool
isValid entry = lower <= count && count <= upper
  where lower = a entry
        upper = b entry
        char  = c entry
        password = p entry
        count = length $ filter (== char) password

File: src/Day2/part1.rs
use crate::day2::util::{Entry, parse_entries};

fn is_valid(entry: &Entry) -> bool {
    let count = entry.p.matches(&entry.c.to_string()).count();
    return entry.a <= count && count <= entry.b;
}

pub fn day2part1(input: &String) -> Option<String> {
    let entries = match parse_entries(input) {
        Some(e) => e,
        None => return None,
    };
    let valid_entries: Vec<&Entry> = entries
        .iter()
        .filter(|e| is_valid(e))
        .collect();
    return Some(valid_entries.len().to_string());
}
